### DISCUSSION TITLE
Snapshot create and play but how ?

### DISCUSSION THREAD

**Discussion by ejonges** [hklages/node-red-contrib-sonos-plus#264] [Q&A] [answered]
Sorry folks for asking this newbie question. I love the SonosPlus function and got it to work with my smart doorbell. Now I`d like to use the snapshot function to restore the state prior to the doorbell ring. So:

1. Sonos plays a music station
2. Doorbell rings
3. Take snapshot via group.create.snap command (I see output via debugger)
4. Play dingdong sample multiple times with some delay (works great)
5. Another couple of seconds delay (works too)
6. Restore snapshot via group.play.snap command

In both 3 & 6 I`ve left the payload to string and blank in the GUI. Step 6 tells me:

24-4-2023 17:26:13[node: play snap](http://192.168.1.61:8123/api/hassio_ingress/tauDcEDClyVKB7-CK6cFWXe6Xo0NJliqBWkIeuKpnt0/#)
msg : string[71]
"group.play.snap:: snapshot (msg.payload) is not object :: Details: none"

So I`m missing something basic but no idea what. Thanks for any assistance.

**Comment by hklages** [comment, marked as answer]
Hi.

You have to store the snapshot in a global or flow variable and then reuse
it in the play snap shot.

That's all. There is an example in the wiki.

**Reply by ejonges** [reply]
Thanks so much; can you guide me to example ? There is a lot of them I saw.

**Comment by ejonges** [comment]
It took some time but found the examples and imported them. Testing is for tomorrow; will let you know the outcome. Thx again.

**Comment by hklages** [comment]
... it is a good exercise :-). Usually Paketes provide some examples in the
library. That should help newbees .

Thanks for feedback and let me know if you need further help.

By the way: Have a look at group.play.notification and player.play.clip.
The first one does the create/restore snapshot internally and the second
one is a new SONOS build in functionality (but only works on one player)

**Comment by ejonges** [comment]
Man, this works like a charm ! Thanks ! Volume level is not restored from the snapshot I believe so still need to look into that. Also I`d like to understand what group.play.notification does but love it !

![image](https://user-images.githubusercontent.com/117987798/234230880-7f40196a-af35-449f-bf9d-7a21a4383d0d.png)

**Comment by hklages** [comment]
Great! Group.play.notification does exactly the same as your flow.

One thing: If you have a playlist playing and then using a doorbell ring
(using  the queue) your flow will not work. For radios and stream it works
fine. Solution is to save and later restore the queue.

Cheers, Henning

**Reply by ejonges** [reply]
Thanks again Henning, I`ll look into the playlist part because that should work too. Mostly we play radio but also Spotify runs frequently. The save & restore queue, can that be integrated in the current flow ? Below the latest & greatest version :-)

![image](https://user-images.githubusercontent.com/117987798/234397777-c8aecd4b-7726-4670-8dda-c6b474faecd3.png)

**Reply by davidhthomsen** [reply]
Nice flow. Is it a possiblillity, that you would share it?

**Reply by ejonges** [reply]
Sure no problem; could you help me out how to do this in the easiest way in Node Red and here on the topic ?

**Comment by hklages** [comment]
... flow looks good. I thing you don't need the group.stop.

For the save queue please have a look at

https://github.com/hklages/node-red-contrib-sonos-plus/wiki/B.-Examples

Example 8.

**Comment by ejonges** [comment]
Just fyi... the group.stop has a reason why I put it in later while testing the flow. Without it, the very last fraction of the played content (so where the snapshot is taken from) is going to the louder volume that I want the Ding Dong to be played in. That`s annoying and I avoid it by first stopping the content that is playing before the ding dong rings at a rather high volume.

**Comment by Travion1976** [comment]
I am trying to get the same implemented, but for some reason it's not working. Any feedback would be great ! 
I have copied the same set up as ejonges.

![Screenshot 2024-07-09 at 11 07 19](https://github.com/hklages/node-red-contrib-sonos-plus/assets/50074166/58cfe119-234a-4f80-a877-ee5c69e3dd42)

When I ring the doorbell this is what I get:
![Screenshot 2024-07-09 at 11 14 50](https://github.com/hklages/node-red-contrib-sonos-plus/assets/50074166/fcb63da2-a638-4cc3-a756-2554250db16a)

There seems to be something wrong when trying to play the snapshot:
msg : string[71]
"group.play.snap:: snapshot (msg.payload) is not object :: Details: none"

This is what I have in my configuration of the group play snapshot:
![Screenshot 2024-07-09 at 11 18 27](https://github.com/hklages/node-red-contrib-sonos-plus/assets/50074166/c742a5b4-0fbe-44d6-8f02-5126775f5e55)

Maybe someone has some input on how to get this solved.
Thanks a lot.

**Comment by hklages** [comment]
That flow uses flow variables. Are you  familiar with flow variables? What
is the content of the flow variables? Can you post?
Grüße Henning Klages

**Reply by Travion1976** [reply]
Hi Henning,
Thank you for your reply. I really don't know what flow variables are, as I am completely new to node red. 
Is this the information when clicking on the individual nodes (Topic and Payload) ?
As in this screenshot? 
Thanks und Grüße
![Screenshot 2024-07-10 at 15 49 42](https://github.com/hklages/node-red-contrib-sonos-plus/assets/50074166/6a1d387c-49aa-4bf8-a376-819b87651f86)

Sebastian

**Comment by hklages** [comment]
Hi Sebastian. 
The logic in that flow is that
1. you create a snapshot of the current goup and store that for later use in a flow variable. 
2. later you restore the data from the flow variable and restore the previous group state. 

Flow and global variable are called "context" see [here](https://nodered.org/docs/user-guide/context)
In the node-red gui, drop down menu, next to the debug messages there are the context data of that flow. You have to use the refresh button to update the data. 

Please check whether the group data are storted there (after having executed the first part of the flow)

**Comment by ejonges** [comment]
Hi guys, it`s been a long time that I tested this flow since I never installed my doorbell in the end and sold it, But.....I just tested it with a dummy trigger and the flow is still alive so let me know if there is anything I can do to help debugging.
